File: vscode/src/layoutCalculator.ts
import { type FieldInfo, effectiveAlignment } from "./structAnalyzer";
import { alignUp } from "./typeSizeTable";

export interface FieldPlacement {
    field: FieldInfo;
    offset: number;
    paddingBefore: number;
}

export interface LayoutResult {
    totalSize: number;
    placements: FieldPlacement[];
    paddingBytes: number;
}

export interface OptimizationResult {
    currentLayout: LayoutResult;
    optimalLayout: LayoutResult;
    bytesSaved: number;
    isAlreadyOptimal: boolean;
    orderedFields: FieldInfo[];
    warnings: string[];
}

export function computeLayout(fields: FieldInfo[], pack: number): LayoutResult {
    const placements: FieldPlacement[] = [];
    let currentOffset = 0;
    let maxAlignment = 0;

    for (const field of fields) {
        const alignment = effectiveAlignment(field, pack);
        const size = field.typeInfo?.size;
        if (!size) continue;
        if (alignment > maxAlignment) maxAlignment = alignment;

        const padding = alignUp(currentOffset, alignment) - currentOffset;
        const alignedOffset = currentOffset + padding;
        placements.push({ field, offset: alignedOffset, paddingBefore: padding });
        currentOffset = alignedOffset + size;
    }

    const structAlignment = pack === 0 ? maxAlignment : Math.min(pack, maxAlignment);
    const trailingPadding = alignUp(currentOffset, structAlignment) - currentOffset;
    const totalSize = currentOffset + trailingPadding;
    const totalPadding = placements.reduce((sum, p) => sum + p.paddingBefore, 0) + trailingPadding;

    return { totalSize, placements, paddingBytes: totalPadding };
}

export function computeOptimalOrder(fields: FieldInfo[], pack: number): OptimizationResult {
    const currentLayout = computeLayout(fields, pack);
    const warnings: string[] = [];

    // Separate fixed vs movable
    const fixed = new Map<number, FieldInfo>();
    const movable: FieldInfo[] = [];

    for (let i = 0; i < fields.length; i++) {
        const field = fields[i];
        if (field.isFixed) {
            fixed.set(i, field);
        } else if (!field.typeInfo) {
            movable.push(field);
            warnings.push(`Unknown type '${field.typeText}' for field '${field.name}' — alignment assumed as 8 (pointer).`);
        } else {
            movable.push(field);
        }
    }

    // Sort movable by alignment DESC, then size DESC
    movable.sort((a, b) => {
        const alignA = effectiveAlignment(a, pack);
        const alignB = effectiveAlignment(b, pack);
        if (alignB !== alignA) return alignB - alignA;
        return (b.typeInfo?.size ?? 0) - (a.typeInfo?.size ?? 0);
    });

    // Reconstruct ordered list, keeping fixed at original positions
    const orderedFields: FieldInfo[] = [];
    let movableIdx = 0;
    for (let i = 0; i < fields.length; i++) {
        if (fixed.has(i)) {
            orderedFields.push(fixed.get(i)!);
        } else if (movableIdx < movable.length) {
            orderedFields.push(movable[movableIdx++]);
        }
    }

    const optimalLayout = computeLayout(orderedFields, pack);
    const bytesSaved = currentLayout.totalSize - optimalLayout.totalSize;
    const isAlreadyOptimal =
        orderedFields.length === fields.length &&
        orderedFields.every((f, i) => f.textRange === fields[i].textRange);

    return { currentLayout, optimalLayout, bytesSaved, isAlreadyOptimal, orderedFields, warnings };
}

export function formatSummary(result: OptimizationResult, i18n: I18n): string {
    if (result.isAlreadyOptimal) {
        return i18n("already_optimal_detail", String(result.currentLayout.totalSize));
    }
    if (result.bytesSaved > 0) {
        const pct = result.currentLayout.totalSize > 0
            ? Math.floor((result.bytesSaved * 100) / result.currentLayout.totalSize)
            : 0;
        const lines = [
            i18n("saves_bytes", String(result.bytesSaved), String(pct)),
            i18n("size_comparison", String(result.currentLayout.totalSize), String(result.optimalLayout.totalSize)),
            i18n("padding_reduced", String(result.currentLayout.paddingBytes), String(result.optimalLayout.paddingBytes)),
        ];
        if (result.warnings.length > 0) {
            lines.push("\n" + i18n("warnings_header"));
            result.warnings.forEach((w) => lines.push("  - " + w));
        }
        return lines.join("\n");
    }
    return i18n("no_improvement_detail", String(result.currentLayout.totalSize));
}

/** Simple i18n function type */
export type I18n = (key: string, ...args: string[]) => string;

File: src/main/kotlin/com/riderpludgemaker/structoptimizer/LayoutCalculator.kt
package com.riderpludgemaker.structoptimizer

/**
 * Computes struct field layout — both current (declaration order) and optimal (descending alignment).
 * Determines whether reordering is beneficial.
 */
object LayoutCalculator {

    data class LayoutResult(
        val totalSize: Int,
        val fieldOffsets: List<FieldPlacement>,
        val paddingBytes: Int,  // total wasted bytes due to alignment
    )

    data class FieldPlacement(
        val field: StructAnalyzer.FieldInfo,
        val offset: Int,     // aligned offset where the field starts
        val paddingBefore: Int, // bytes of padding inserted before this field
    )

    data class OptimizationResult(
        val currentLayout: LayoutResult,
        val optimalLayout: LayoutResult,
        val bytesSaved: Int,
        val isAlreadyOptimal: Boolean,
        val orderedFields: List<StructAnalyzer.FieldInfo>, // fields in optimal order
        val warnings: List<String>,
    )

    /**
     * Computes the layout for fields in a given order.
     * Returns offsets, padding, and total struct size.
     */
    fun computeLayout(
        fields: List<StructAnalyzer.FieldInfo>,
        pack: Int,
    ): LayoutResult {
        val placements = mutableListOf<FieldPlacement>()
        var currentOffset = 0
        var maxAlignment = 0

        for (field in fields) {
            val alignment = StructAnalyzer.effectiveAlignment(field, pack)
            val size = field.typeInfo?.size ?: continue

            if (alignment > maxAlignment) maxAlignment = alignment

            val padding = TypeSizeResolver.alignUp(currentOffset, alignment) - currentOffset
            val alignedOffset = currentOffset + padding

            placements.add(FieldPlacement(field, alignedOffset, padding))
            currentOffset = alignedOffset + size
        }

        // Total struct size must be a multiple of maxAlignment (for arrays)
        val structAlignment = if (pack == 0) maxAlignment else minOf(pack, maxAlignment)
        val trailingPadding = TypeSizeResolver.alignUp(currentOffset, structAlignment) - currentOffset
        val totalSize = currentOffset + trailingPadding

        // Count total padding
        val totalPadding = placements.sumOf { it.paddingBefore } + trailingPadding

        return LayoutResult(
            totalSize = totalSize,
            fieldOffsets = placements,
            paddingBytes = totalPadding,
        )
    }

    /**
     * Determines the optimal field order to minimize padding.
     *
     * Strategy: sort fields by effective alignment descending, then by size descending.
     * Fields with unresolved types are placed at the end (largest alignment conservative).
     *
     * Fields that should NOT be reordered (fixed buffers, etc.) are kept in their relative
     * positions among the movable fields.
     */
    fun computeOptimalOrder(
        fields: List<StructAnalyzer.FieldInfo>,
        pack: Int,
    ): OptimizationResult {
        val currentLayout = computeLayout(fields, pack)

        val warnings = mutableListOf<String>()

        // Separate fields into movable and fixed
        val movable = mutableListOf<StructAnalyzer.FieldInfo>()
        val fixed = mutableListOf<Pair<Int, StructAnalyzer.FieldInfo>>() // (originalIndex, field)

        for ((i, field) in fields.withIndex()) {
            if (field.isFixed) {
                fixed.add(i to field)
            } else if (field.typeInfo == null) {
                // Unknown type — keep in place with a warning
                movable.add(field)
                warnings.add(Messages.get("unknown_type", field.typeText, field.fieldName))
            } else {
                movable.add(field)
            }
        }

        // Sort movable fields by effective alignment DESC, then size DESC
        val sortedMovable = movable.sortedWith(
            compareByDescending<StructAnalyzer.FieldInfo> { field ->
                StructAnalyzer.effectiveAlignment(field, pack)
            }.thenByDescending { field ->
                field.typeInfo?.size ?: 0
            }
        )

        // Place fixed fields in their original positions, based on original indices
        val orderedFields = mutableListOf<StructAnalyzer.FieldInfo>()
        var movableIdx = 0
        val fixedSet = fixed.map { it.first }.toSet()

        for (i in fields.indices) {
            if (i in fixedSet) {
                // Insert the fixed field at its original position
                val fixedField = fixed.first { it.first == i }.second
                orderedFields.add(fixedField)
            } else {
                // Take the next movable field from the sorted list
                if (movableIdx < sortedMovable.size) {
                    orderedFields.add(sortedMovable[movableIdx])
                    movableIdx++
                }
            }
        }

        // Compute optimal layout
        val optimalLayout = computeLayout(orderedFields, pack)
        val bytesSaved = currentLayout.totalSize - optimalLayout.totalSize
        val isAlreadyOptimal = orderedFields.map { it.textRange }.toList() ==
                               fields.map { it.textRange }.toList()

        return OptimizationResult(
            currentLayout = currentLayout,
            optimalLayout = optimalLayout,
            bytesSaved = bytesSaved,
            isAlreadyOptimal = isAlreadyOptimal,
            orderedFields = orderedFields,
            warnings = warnings,
        )
    }

    /**
     * Generates a human-readable summary of the optimization.
     */
    fun formatSummary(result: OptimizationResult): String {
        val sb = StringBuilder()
        if (result.isAlreadyOptimal) {
            sb.append(Messages.get("already_optimal_detail", result.currentLayout.totalSize))
        } else if (result.bytesSaved > 0) {
            val pct = if (result.currentLayout.totalSize > 0) {
                (result.bytesSaved * 100) / result.currentLayout.totalSize
            } else 0
            sb.appendLine(Messages.get("saves_bytes", result.bytesSaved, pct))
            sb.appendLine(Messages.get("size_comparison", result.currentLayout.totalSize, result.optimalLayout.totalSize))
            sb.appendLine(Messages.get("padding_reduced", result.currentLayout.paddingBytes, result.optimalLayout.paddingBytes))
        } else {
            sb.append(Messages.get("no_improvement_detail", result.currentLayout.totalSize))
        }
        if (result.warnings.isNotEmpty()) {
            sb.appendLine("\n" + Messages.get("warnings_header"))
            result.warnings.forEach { sb.appendLine("  - $it") }
        }
        return sb.toString()
    }
}
